File: api/bin/lib/models/asset.dart
/// Asset model — mirrors the `Asset` interface in src/app/lib/types.ts.
///
/// Two factory paths:
///   - `Asset.fromJson(Map)` — used by the HTTP layer; expects ISO date
///     strings, plain JSON numbers/booleans.
///   - `Asset.fromRow(Map)`  — used by repos; expects DateTime objects and
///     `num` from the postgres package.
///
/// Run `dart run build_runner build` to regenerate `.freezed.dart` / `.g.dart`.
library;

import 'package:freezed_annotation/freezed_annotation.dart';

import 'json_helpers.dart';

part 'asset.freezed.dart';
part 'asset.g.dart';

@Freezed()
abstract class Asset with _$Asset {
  const Asset._();

  const factory Asset({
    required String id,
    required String companyId,
    required String mdmSource,
    required String modelName,
    required String assetType,
    required String conditionGrade,
    required int    quantity,
    @IsoDateTimeConverter()      required DateTime createdAt,
    @IsoDateTimeConverter()      required DateTime updatedAt,
    String?  serialNumber,
    String?  reasonForOffload,
    @NullableDateOnlyConverter()      DateTime? purchaseDate,
    double?  originalPurchasePrice,
    String?  osVersion,
    double?  batteryHealthPct,
    int?     batteryCycles,
    String?  complianceState,
    String?  assignedUser,
    String?  department,
    String?  costCenter,
    @NullableIsoDateTimeConverter()   DateTime? lastMdmSync,
    double?  cpuScore,
    double?  ramGb,
    double?  storageGb,
    // ── Marketplace device-spec fields (buyer-facing listing detail) ────────
    String?  manufacturer,
    String?  modelNumber,
    int?     yearOfManufacture,
    String?  functionalStatus,
    String?  knownDefects,
    String?  dataWipeStatus,
    String?  warrantyStatus,
    @NullableIsoDateTimeConverter()   DateTime? warrantyExpiration,
    String?  includedAccessories,
    String?  shipsFromLocation,
    String?  cpuModel,
    int?     cpuCores,
    double?  cpuSpeedGhz,
    String?  ramType,
    String?  storageType,
    String?  gpuModel,
    double?  screenSizeIn,
    String?  screenResolution,
    bool?    touchscreen,
    String?  formFactor,
    int?     powerSupplyWatts,
    String?  panelType,
    int?     refreshRateHz,
    String?  ports,
    int?     portCount,
    String?  throughput,
    bool?    managed,
    bool?    carrierLocked,
  }) = _Asset;

  factory Asset.fromJson(Map<String, dynamic> json) => _$AssetFromJson(json);

  /// Build directly from a Postgres row map. Date columns come back as
  /// `DateTime` already; numerics as `num`.
  factory Asset.fromRow(Map<String, dynamic> row) => Asset(
        id:                    row['id']                    as String,
        companyId:             row['company_id']            as String,
        mdmSource:             enumStr(row['mdm_source']),
        modelName:             row['model_name']            as String,
        assetType:             enumStr(row['asset_type']),
        conditionGrade:        enumStr(row['condition_grade']),
        quantity:              numToIntOrNull(row['quantity']) ?? 1,
        createdAt:             row['created_at']            as DateTime,
        updatedAt:             row['updated_at']            as DateTime,
        serialNumber:          row['serial_number']         as String?,
        reasonForOffload:      row['reason_for_offload']    as String?,
        purchaseDate:          row['purchase_date']         as DateTime?,
        originalPurchasePrice: numToDoubleOrNull(row['original_purchase_price']),
        osVersion:             row['os_version']            as String?,
        batteryHealthPct:      numToDoubleOrNull(row['battery_health_pct']),
        batteryCycles:         numToIntOrNull(row['battery_cycles']),
        complianceState:       row['compliance_state']      as String?,
        assignedUser:          row['assigned_user']         as String?,
        department:            row['department']            as String?,
        costCenter:            row['cost_center']           as String?,
        lastMdmSync:           row['last_mdm_sync']         as DateTime?,
        cpuScore:              numToDoubleOrNull(row['cpu_score']),
        ramGb:                 numToDoubleOrNull(row['ram_gb']),
        storageGb:             numToDoubleOrNull(row['storage_gb']),
        manufacturer:          row['manufacturer']          as String?,
        modelNumber:           row['model_number']          as String?,
        yearOfManufacture:     numToIntOrNull(row['year_of_manufacture']),
        functionalStatus:      enumStrOrNull(row['functional_status']),
        knownDefects:          row['known_defects']         as String?,
        dataWipeStatus:        enumStrOrNull(row['data_wipe_status']),
        warrantyStatus:        enumStrOrNull(row['warranty_status']),
        warrantyExpiration:    row['warranty_expiration']   as DateTime?,
        includedAccessories:   row['included_accessories']  as String?,
        shipsFromLocation:     row['ships_from_location']   as String?,
        cpuModel:              row['cpu_model']             as String?,
        cpuCores:              numToIntOrNull(row['cpu_cores']),
        cpuSpeedGhz:           numToDoubleOrNull(row['cpu_speed_ghz']),
        ramType:               row['ram_type']              as String?,
        storageType:           enumStrOrNull(row['storage_type']),
        gpuModel:              row['gpu_model']             as String?,
        screenSizeIn:          numToDoubleOrNull(row['screen_size_in']),
        screenResolution:      row['screen_resolution']     as String?,
        touchscreen:           row['touchscreen']           as bool?,
        formFactor:            row['form_factor']           as String?,
        powerSupplyWatts:      numToIntOrNull(row['power_supply_watts']),
        panelType:             row['panel_type']            as String?,
        refreshRateHz:         numToIntOrNull(row['refresh_rate_hz']),
        ports:                 row['ports']                 as String?,
        portCount:             numToIntOrNull(row['port_count']),
        throughput:            row['throughput']            as String?,
        managed:               row['managed']               as bool?,
        carrierLocked:         row['carrier_locked']        as bool?,
      );
}

File: guildmark/src/app/models/asset.ts
// ---------------------------------------------------------------------------
// Assets
// ---------------------------------------------------------------------------

export type AssetType =
  | "laptop" | "desktop" | "server" | "phone"
  | "tablet" | "networking" | "monitor" | "software" | "license" | "other";

export type ConditionGrade = "A" | "B" | "C";

export type MdmSource = "manual" | "jamf_pro" | "jamf_school" | "intune";

export type FunctionalStatus =
  | "fully_functional" | "functional_with_issues" | "for_parts";
export type DataWipeStatus = "not_wiped" | "wiped" | "certified";
export type WarrantyStatus = "none" | "active" | "expired";
export type StorageType = "ssd" | "nvme" | "hdd" | "emmc" | "other";

export interface Asset {
  id:                      string;
  company_id:              string;
  mdm_source:              MdmSource;
  serial_number:           string | null;
  model_name:              string;
  asset_type:              AssetType;
  condition_grade:         ConditionGrade;
  quantity:                number;
  reason_for_offload:      string | null;
  purchase_date:           string | null;
  original_purchase_price: number | null;
  os_version:              string | null;
  battery_health_pct:      number | null;
  battery_cycles:          number | null;
  compliance_state:        string | null;
  assigned_user:           string | null;
  department:              string | null;
  cost_center:             string | null;
  last_mdm_sync:           string | null;
  // Spec metadata (sourced from MDM where available; manual otherwise)
  cpu_score:               number | null;
  ram_gb:                  number | null;
  storage_gb:              number | null;
  // Marketplace device-spec detail (manual entry / listing form)
  manufacturer?:           string;
  model_number?:           string;
  year_of_manufacture?:    number | null;
  functional_status?:      FunctionalStatus;
  known_defects?:          string;
  data_wipe_status?:       DataWipeStatus;
  warranty_status?:        WarrantyStatus;
  warranty_expiration?:    string | null;
  included_accessories?:   string;
  ships_from_location?:    string;
  cpu_model?:              string;
  cpu_cores?:              number | null;
  cpu_speed_ghz?:          number | null;
  ram_type?:               string;
  storage_type?:           StorageType;
  gpu_model?:              string;
  screen_size_in?:         number | null;
  screen_resolution?:      string;
  touchscreen?:            boolean;
  form_factor?:            string;
  power_supply_watts?:     number | null;
  panel_type?:             string;
  refresh_rate_hz?:        number | null;
  ports?:                  string;
  port_count?:             number | null;
  throughput?:             string;
  managed?:                boolean;
  carrier_locked?:         boolean;
  created_at:              string;
  updated_at:              string;
}
